core/include/detray/core/detail/context_aware_store.hpp

```diff
@@ -0,0 +1,360 @@
+              std::enable_if_t<detail::is_device_view_v<container_view_t>,
+                               bool> = true>
+    DETRAY_HOST_DEVICE context_aware_store(container_view_t& view)
+        : m_container(view) {}
+
+    /// @returns a pointer to the underlying container - const
+    DETRAY_HOST_DEVICE
+    constexpr auto data() const noexcept -> const base_type* {
+        return &m_container;
+    }
+
+    /// @returns a pointer to the underlying container - non-const
+    DETRAY_HOST_DEVICE
+    constexpr auto data() noexcept -> base_type* { return &m_container; }
+
+    /// @returns the size of the underlying container
+    DETRAY_HOST_DEVICE
+    constexpr auto size(const context_type& ctx) const noexcept -> dindex {
+        return static_cast<dindex>(m_container.at(ctx.get()).size());
+    }
+
+    /// @returns true if the underlying container is empty
+    DETRAY_HOST_DEVICE
+    constexpr auto empty(const context_type& ctx) const noexcept -> bool {
+        return m_container.at(ctx.get()).empty();
+    }
+
+    /// @returns the collections iterator at the start position.
+    DETRAY_HOST_DEVICE
+    constexpr auto begin(const context_type& ctx) {
+        return m_container.at(ctx.get()).begin();
+    }
+
+    /// @returns the collections iterator sentinel.
+    DETRAY_HOST_DEVICE
+    constexpr auto end(const context_type& ctx) {
+        return m_container.at(ctx.get()).end();
+    }
+
+    /// @returns access to the second-level vector in underlying container -
+    /// const
+    DETRAY_HOST_DEVICE
+    auto get(const context_type& ctx) const noexcept -> const vector_type& {
+        return m_container.at(ctx.get());
+    }
+
+    /// @returns access to the second-level vector in underlying container -
+    /// non-const
+    DETRAY_HOST_DEVICE
+    auto get(const context_type& ctx) noexcept -> vector_type& {
+        return m_container.at(ctx.get());
+    }
+
+    /// @returns access to the underlying container - const
+    DETRAY_HOST_DEVICE
+    auto get() const noexcept -> const base_type& { return m_container; }
+
+    /// @returns access to the underlying container - non-const
+    DETRAY_HOST_DEVICE
+    auto get() noexcept -> base_type& { return m_container; }
+
+    // TODO: does it make sense to introduce the operator[] for this store?
+    /// Elementwise access. Needs @c operator[] for storage type - non-const
+    /*
+    DETRAY_HOST_DEVICE
+    constexpr decltype(auto) operator[](const dindex i) {
+      return m_container[i];
+    }
+
+    /// Elementwise access. Needs @c operator[] for storage type - const
+    DETRAY_HOST_DEVICE
+    constexpr decltype(auto) operator[](const dindex i) const {
+      return m_container[i];
+    }
+    */
+
+    /// @returns context based access to an element (also range checked)
+    DETRAY_HOST_DEVICE
+    constexpr auto at(const dindex i, const context_type& ctx) const noexcept
+        -> const T& {
+        return m_container.at(ctx.get()).at(i);
```

Comment:
Isn't using ```at``` slower for host jagged vectors, because of the bounds checking and the possibility that it throws an exception? Or does it just not make a big difference?

Reply:
As far as I know, the bound checking only happens for debug builds.

Reply:
In device code I think you are correct, but what about host code? ```std::vector<>::at``` can throw an exception, too,  if the index is out of bounds. So, the ```noexcept``` should probably be removed. Anyway, I don't think it will make a big performance difference

Reply:
OK. I didn't know that `at` of std::vector does the bound checking. But I also don't think it will make a big difference for host code